doc/api/quic.md

```diff
@@ -0,0 +1,1344 @@
+# QUIC
+
+## Overview
+
+The `net/quic` module provides client and server implementations of the QUIC
+and HTTP/3 protocols.
+
+```mjs
+import {
+  Endpoint,
+} from 'node:net/quic';
+```
+
+```cjs
+const {
+  Endpoint,
+} = require('net/quic');
+```
+
+The {quic.Endpoint} object is the primary entrypoint for creating
+both clients and servers.
+
+### A simple QUIC Server
+
+```mjs
+import {
+  Endpoint,
+} from 'node:net/quic';
+
+const server = new Endpoint();
+
+server.onsession = ({ session }) => {
+  session.onstream = ({ stream, respondWith }) => {
+    respondWith({ body: 'Hello World' });
+  };
+};
+
+server.listen({ alpn: 'xyz' });
+```
+
+```cjs
+const {
+  Endpoint,
+} = require('net/quic');
+
+const server = new Endpoint();
+
+server.onsession = ({ session }) => {
+  session.onstream = ({ stream, respondWith }) => {
+    respondWith({ body: 'Hello World' });
+  };
+};
+
+server.listen({ alpn: 'xyz' });
+```
+
+### A simple QUIC Client
+
+```mjs
+import {
+  Endpoint,
+} from 'node:net/quic';
+
+const client = new Endpoint();
+const session = client.connect({
+  address: '123.123.123.123',
+  port: 1234,
+}, { alpn: 'xyz' });
+const stream = session.open({ body: 'Hello World' });
+console.log(await stream.text());
+```
+
+```cjs
+const {
+  Endpoint,
+} = require('net/quic');
+
+async function sayHi() {
+  const client = new Endpoint();
+  const session = client.connect({
+    address: '123.123.123.123',
+    port: 1234,
+  }, { alpn: 'xyz' });
+  const stream = session.open({ body: 'Hello World' });
+  console.log(await stream.text());
+}
+
+sayHi();
+```
+
+### HTTP/3
+
+#### A simple HTTP/3 Server
+
+```mjs
+import {
+  Endpoint,
+} from 'node:net/quic';
+
+import { stdout } from 'node:process';
+
+const endpoint = new Endpoint({ address: { port: 12345 } });
+
+endpoint.onsession = ({ session }) => {
+  session.onstream = async ({ stream, respondWith }) => {
+    respondWith({
+      headers: { 'content-type': 'text/plain' },
+      body: 'right back at you',
+    });
+
+    console.log(await stream.headers);
+
+    stream.readableNodeStream().pipe(stdout);
+  };
+};
+
+endpoint.listen({
+  secure: {
+    key: getTLSKeySomehow(),
+    cert: getTLSCertSomehow(),
+  },
+});
+```
+
+#### A simple HTTP/3 Client
+
+```mjs
+import {
+  Endpoint,
+} from 'node:net/quic';
+
+import { stdout } from 'node:process';
+
+const client = new Endpoint();
+
+const req = client.connect({
+  address: '127.0.0.1',
+  port: 12345
+}, {
+  hostname: 'localhost',
+});
+
+await req.handshake;
+
+const stream = req.open({
+  headers: { ':method': 'POST' },
+  body: 'hello there',
+});
+
+console.log(await stream.headers);
+
+stream.readableNodeStream().pipe(stdout);
+
+await client.close();
```

Comment:
@mcollina: Can we somehow unify this with undici?

Reply:
```js
const client = new Client({
  address: '127.0.0.1',
  port: 12345
}, {
  hostname: 'localhost',
});

const { body, headers } = await client.request({
  headers: { ':method': 'POST' },
  body: 'hello there',
});

stream.readableNodeStream().pipe(stdout);

await client.close();
```

Reply:
We would have to change undici to return `body` as an object with `.text()`, `readableNodeSream()`, `readableWebStream()` etc. But I'm open for that.

Reply:
Similarly to the server side. I don't want to rename `open()` to `request()` because the `open()` event is also available on the server-side (the server can initiate unidirectional and bidirectional streams back to the client at any time). The `session.open()` method is the same on both sides. And again, the API here is generic for all QUIC apps, not just HTTP, and should not special case HTTP/3 too much.